Culture settings

Skrypt opened the pull request:
Please refer to issue #434 for previous discussion.  

![image](https://user-images.githubusercontent.com/3228637/42112746-e1aaacb4-7bb6-11e8-8c11-32717db081bc.png)
![image](https://user-images.githubusercontent.com/3228637/42118008-22737ffc-7bce-11e8-8f81-839cf76cb388.png)
![image](https://user-images.githubusercontent.com/3228637/42427583-45da0e8a-82fd-11e8-84f0-10d67080cf35.png)


TODO: 

- ~~Investigate the use of the app.UseRequestLocalization() middleware vs using our own Scoped middleware through the use of ILocalCulture. The default middleware will alter current thread CultureInfo.CurrentUICulture and CultureInfo.CurrentCulture. Since in 01 we specifically wanted to avoid this ; second solution would be more apropriate. First solution `app.UseRequestLocalization() middleware` is currently implemented in OrchardCore.Localization.Startup.~~
- ~~Complete implementation of CultureManager and other places where we need to get Culture from the IHttpContextAccessor. We had in 01 a IWorkContextAccessor, not sure, as of now, if it's needed in Core and if not then what replaces it.~~
- Add culture selector in Setup. By default we used to install "en-US" culture which is replaced by InvariantCulture. Though we should be able to set the default culture from the Setup as a Recipe step.
- ~~Validate that custom cultures works properly.~~
- ~~Test x-plat compatibility~~
- ~~Use a single database row in Document table for storing supported cultures.~~

Related Issues : 

#434 #2073 #736 #1462 #1740

Skrypt commented:
CI error reported is : 

```
Updating build cache...
Cache entry not found: C:\projects\orchard2\node_modules
Cache 'C:\Users\appveyor\AppData\Local\NuGet\v3-cache' - Calculating checksum...Unable to connect to the remote server
```

Though it's still not working properly since I merged latest dev branch. I get an error with : 

```
InvalidOperationException: Unable to resolve service for type 'OrchardCore.Localization.Services.ICultureManager' while attempting to activate 'OrchardCore.Settings.Controllers.AdminController'.
```

There is still some work to do on that PR. I'm getting on it.

agriffard commented:
If I start from a clean setup, there is an error when I open the General settings:
`Microsoft.Data.Sqlite.SqliteException: SQLite Error 1: 'no such table: CultureIndex'.`

I guess the OrchardCore.Localization module needs to be included in the recipe and the OrchardCore.Settings needs a dependency on it, or else the migration is not executed.

Skrypt commented:
Yes there is still a bug with that though we need to decide if we want to always have that module ON like we did in 01 since it was part of the Orchard.Core project. Else I can check if I can create 2 separate features for PO localization and Localization settings.

jtkech commented:
I made some suggestions, see #2076.

Some suggestions

- Migration for `CultureIndex` is done in `OC.Settings` so that it doesn't need a dependency on `OC.Localization`.

- As it is used we don't need to register `DefaultRequestCultureProvider`, and it is defined in `OC.Localization.Core` so that `OC.Localization` doesn't need to reference `OC.Settings`.

- In the startup of `OC.Localization`, use the new helpers `SetDefaultCulture`, `AddSupportedCultures` and `AddSupportedUICultures`.

- And other minor updates.

A few others to come, nothing has been tested yet ;)

Skrypt commented:
I merged your PR on this PR branch. The only issue left is that when we set a default culture in the admin it should change the default culture but it doesn't since it retreives that value from the DefaultRequestCultureProvider that is never updated once we updated the Database value.

Skrypt commented:
Forget what I said in my previous post everything is working fine. I probably had a compilation issue last night after merging @jtkech PR. I can't set this PR as ready because I still need to make the DB records to be set in a single row like @sebastienros asked last week. Though feel free to test it and report any bugs.

jtkech commented:
Okay, i understand the last concern. Do you think we really need to find a solution?

Right now, you just have to disable and re-enable `Localization`.
Or enable `Tenants` and then reload the given tenant.

I will take a look, maybe we could automatically reload the tenant as we do when changing a feature.

Skrypt commented on src/OrchardCore/OrchardCore.Modules/Services/LocalCulture.cs:
```diff
@@ -26,7 +26,7 @@ public Task<CultureInfo> GetLocalCultureAsync()
         public bool IsLocalizationEnabled()
         {
-            return _httpContextAccessor.HttpContext.Features.Get<IRequestCultureFeature>() == null ? false : true;
+            return _httpContextAccessor.HttpContext.Features.Get<IRequestCultureFeature>() != null;
```
lol @ me

jtkech replied:
;)

Skrypt commented:
I don't know for now. From reading the comments they seem to not have made that available by design. Something related to culture cache.

Skrypt commented:
I've been thinking about your suggestion for reloading tenants after adding a new culture. I'd say that if we add this feature then maybe we should add it as optional. Maybe something like a pop-up that says "reload tenant?" or maybe a simply a checkbox/button in the general settings view. But as optional because I don't like the idea of having tenant reloads on every new culture addition or removal.

jtkech commented:
Yes i agree.

In the last commit, just to show.

- Introduce `RouteDataRequestCultureProvider`.

- Do all the stuff in the startup `Configure()`.

I will create a custom interface to wrap the request culture providers so that each provider will have an `Order` property. This way, any module will be able to register a provider with a given priority. Hmm, and then we could register again the `DefaultRequestCultureProvider` through the `Settings` module.

Need to leave, i will continue this night.

Skrypt commented:
Added DataAnnotation localization

Skrypt commented on src/OrchardCore/OrchardCore.DisplayManagement/Shapes/DateTimeShapes.cs:
```diff
@@ -5,24 +5,28 @@
     public class DateTimeShapes : IShapeAttributeProvider
     {
-        private const string LongDateTimeFormat = "dddd, MMMM d, yyyy h:mm:ss tt";
+        //private const string LongDateTimeFormat = "dddd, MMMM d, yyyy h:mm:ss tt";
```
@sebastienros Maybe you could tell me why we we're using a string like this here. When localizing this not works so well since the order of day, month, year can change per culture. Maybe the intent was to get more details on the date itself by adding "tt" in the end. Else, when removing this, dates are getting properly localized.

Skrypt commented on src/OrchardCore/OrchardCore.DisplayManagement/Shapes/DateTimeShapes.cs:
```diff
@@ -81,13 +85,15 @@ public IHtmlContent TimeSpan(IHtmlHelper Html, DateTime? Utc, DateTime? Origin)
 
-            if (Format == null)
-            {
-                Format = T[LongDateTimeFormat, LongDateTimeFormat, 0].Value;
```
Why using pluralization here?

Skrypt commented:
It's cleaned up though I found that there's something strange with DateTime string format and localization with Liquid files at least. Not sure if it's related with Liquid I need to debug and see.

Skrypt commented:
Ok I removed the AcceptLanguageHeaderRequestCultureProvider for now. I tested it and personnally I don't like how it behaves for multiple reasons. For example I set my default culture to fr-FR and I added a en-US supportedCulture. Then on my browser I set the default culture to en-US and remove all the other cultures. I get the website as en-US instead of fr-FR. The Provider works as intended because my browser doesn't support fr-FR, but I won't let Orchard install that culture just because I'm supporting it. So, globally, I don't like how this Provider works.

Skrypt commented:
I've started looking at the RouteDataRequestCultureProvider last night for having routes that uses a culture param. Also, the DefaultRequestCultureProvider that we removed was coming from an example from .NET Core documentation, so using custom culture providers is encouraged. That custom provider was simply changing current culture based on the default culture selected in the admin settings and it was doing it without needing to reset the tenant "it was changing the current culture at runtime¨. Now we removed it in favor of resetting the tenant everytime we use the save button in the default culture settings view.

pgy866 commented:
I see that the "Cultures this site supports" list is empty by default.
Should I put the default Default Culture,
Default is added to the "Cultures this site supports" list.

![1](https://user-images.githubusercontent.com/28694676/43906534-10e5ba3e-9c26-11e8-9769-4d277a2d0064.gif)

pgy866 commented:
Sorry, if there is only one culture by default, it seems that adding to the list is superfluous.
